### PR TITLE
Integration with Project Reactor

### PR DESCRIPTION
**Pull request by lucasvalenteds**
This PR adds a new module that allows users to use [Mono](https://projectreactor.io/docs/core/milestone/reference/docs/index.html#mono) class from [Project Reactor](https://github.com/reactor/reactor-core) in projects written using reactive programming style targeting JVM.

It follows the same conventions used throughout the project for naming and error handling. There are functions `monoOfResponse`, `monoOf<Type>` and `monoOfResult<Type>` as in RxJava and Coroutines modules.

Please, let me know if you have any comments!

### CODE REVIEW

File: fuel-reactor/build.gradle [hunk 1]
```diff
@@ -0,0 +1,43 @@
+apply plugin: 'kotlin'
```
Comment (kittinunf) [inline fuel-reactor/build.gradle]: I think this needs to be modified a bit after the Kotlin-dsl made it into master #423, right?

Comment (iNoles) [reply fuel-reactor/build.gradle]:
```
plugins {
   kotlin("jvm") 
}
```

Comment (lucasvalenteds) [reply fuel-reactor/build.gradle]: Changed in e1180d2d05b6ad2dc8e2ed0c8fcc6165fc57a7f2

File: fuel-reactor/src/main/kotlin/com/github/kittinunf/fuel/reactor/Reactor.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+fun Request.monoOfBytes(): Mono<ByteArray> =
+    monoOfResultFold(ByteArrayDeserializer())
+
+fun Request.monoOfString(charset: Charset = Charsets.UTF_8): Mono<String> =
+    monoOfResultFold(StringDeserializer(charset))
+
+fun <T : Any> Request.monoOfObject(mapper: Deserializable<T>): Mono<T> =
+    monoOfResultFold(mapper)
+
+private fun <T : Any> Request.monoOfResultUnFolded(mapper: Deserializable<T>): Mono<Result<T, FuelError>> =
```
Comment (kittinunf) [inline fuel-reactor/src/main/kotlin/com/github/kittinunf/fuel/reactor/Reactor.kt]: Can this method and `monoOfResultFold` use one private method? Practically, `monoOfResultFold` is a map version of  `monoOfResultUnFolded`?

File: fuel-reactor/src/main/kotlin/com/github/kittinunf/fuel/reactor/Reactor.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import java.nio.charset.Charset
+
+private fun <T : Any> Request.monoOfResultFold(mapper: Deserializable<T>): Mono<T> =
+    Mono.create<T> { sink ->
+        sink.onCancel(this::cancel)
+        val (_, _, result) = response(mapper)
+        result.fold(sink::success, sink::error)
+    }
+
+fun Request.monoOfBytes(): Mono<ByteArray> =
```
Comment (kittinunf) [inline fuel-reactor/src/main/kotlin/com/github/kittinunf/fuel/reactor/Reactor.kt]: [Q] Shall we omit the `Of` here so it kinda goes similar to the `responseXXX` version? 

`responseBytes` => `monoBytes`
`responseString` => `monoString`
`responseObject` => `monoObject`

Comment (lucasvalenteds) [reply fuel-reactor/src/main/kotlin/com/github/kittinunf/fuel/reactor/Reactor.kt]: I prefer `monoOf`, but I agree that using `mono` only can be better for consistency. Changed in f669d3c.

File: fuel-reactor/src/main/kotlin/com/github/kittinunf/fuel/reactor/Reactor.kt (hunk 2 above)
Comment (lucasvalenteds) [reply fuel-reactor/src/main/kotlin/com/github/kittinunf/fuel/reactor/Reactor.kt]: Changed in e9add41.

File: fuel-reactor/src/main/kotlin/com/github/kittinunf/fuel/reactor/Reactor.kt (hunk 3 above)
Comment (SleeplessByte) [reply fuel-reactor/src/main/kotlin/com/github/kittinunf/fuel/reactor/Reactor.kt]: `monoOf<T>` would still work, and agreed with all the other comments 🗡

**Comment by kittinunf** [comment]
I think it looks great to me. Regarding the test, we might need @SleeplessByte to shed some shiny light ✨ ⚡️ to make it inline with the mock server thingy though. 

Also, it could be done in the follow-up PR.

**Comment by iNoles** [comment]
ReadmeSample**.kt should not be needed for this project.

**Comment by SleeplessByte** [comment]
@kittinunf if you take a look at #444, it should be clear how to use mock-server for this. In general:

1. add the depedency if it's not there yet
2. remove the base path
3. each test do the following:
  a. `@Before`: create a mock server and start it
  b. `mock.chain(request = <request you are expecting>, response = <response you want to>` 
  c. `Fuel.method(mock.path("the path"))`
  d. `@After`: shutdown mock server

You can usually use `mock.reflect()` which will simply give back the request with some body that holds what you sent.

**Comment by lucasvalenteds** [comment]
Thank you for the feedback @kittinunf, @iNoles and @SleeplessByte. The changes were addressed in 205f18a and b927e40.

Let me know if there is anything else.

**Comment by kittinunf** [comment]
Looks great for me.

**Pull request merged**